presto-main/src/main/java/io/prestosql/sql/planner/RuleStatsRecorder.java

```diff
@@ -36,6 +41,7 @@
 public class RuleStatsRecorder
 {
     private final Map<Class<?>, RuleStats> stats = new HashMap<>();
+    private final Cache<QueryId, Map<Class<?>, QueryRuleStats>> queryRuleStats = CacheBuilder.newBuilder().build();
```

Comment:
I'm not sure about this approach. It looks like a hack around plan optimizers being created globally (once for all queries).
Maybe we should change `io.prestosql.sql.planner.PlanOptimizers#get` so that it creates optimizers with rule stats for particular query?
WDYT: @martint ?

Reply:
Yeah, this is problematic. A better approach would be to pass some kind of context object to optimizers (similar to WarningCollector) that can be used to track stats for a specific query. This is related to something I was looking into a while ago -- I wanted to see if we could extend EXPLAIN to be able to describe the sequence of optimization actions for a given query. E.g., something like `EXPLAIN (TYPE OPTIMIZATION)`. It would require a similar mechanism.

Reply:
> Yeah, this is problematic. A better approach would be to pass some kind of context object to optimizers (similar to WarningCollector) that can be used to track stats for a specific query. 

That sounds like a good idea.